distributed/tests/test_worker.py

```diff
@@ -2514,3 +2514,91 @@ def __call__(self, *args, **kwargs):
 
     finally:
         threadpool.shutdown()
+
+
+@gen_cluster(client=True)
+async def test_gather_dep_exception_one_task(c, s, a, b):
+    """Ensure an exception in a single task does not tear down an entire batch of gather_dep
+
+
+    See also https://github.com/dask/distributed/issues/5152
+    See also test_gather_dep_exception_one_task_2
+    """
+    fut = c.submit(inc, 1, workers=[a.address], key="f1")
+    fut2 = c.submit(inc, 2, workers=[a.address], key="f2")
+    fut3 = c.submit(inc, 3, workers=[a.address], key="f3")
+
+    def sink(a, b, *args):
+        return a + b
+
+    res1 = c.submit(sink, fut, fut2, fut3, workers=[b.address])
+    res2 = c.submit(sink, fut, fut2, workers=[b.address])
+
+    import distributed.worker
+
+    # We'll need to modify the tasks after we start the connect but before the
+    # response is handled
+    with mock.patch.object(
+        distributed.worker,
+        "get_data_from_worker",
+        side_effect=distributed.worker.get_data_from_worker,
+    ) as spy:
+
+        while not spy.call_count:
+            await asyncio.sleep(0)
```

Comment:
It's unclear to me why this section is needed / what it's trying to accomplish. It'd be great is we could avoid mocking, though I understand sometimes it's necessary

Reply:
Well, this condition is extremely timing sensitive and I need to manipulate the state while the worker is handling that coroutine. This mock doesn't change any behaviour but rather observes and exits exactly when that happens. This is also why the asyncio sleep is exactly zero because I want to hit the exact loop iteration after the get_data_from_worker is entered.

There are a few other options we can leverage here. Since this is a kind of test I'll likely write in the future again, and have in the past, I provided three flavours of tests we can use to test this kind of behaviour. All of them require a rather implementation specific knowledge. However, typing these, I slightly favour the flavour with the Comm/Event/Queue mechansim since this does not rely on specific functions to be mocked but rather conecptionally intercepts network traffic in a reliable manner. This is something I could factor out in utils_test and we could reuse this.

I am not favouring the proc suspension version since this is very unreliable. I wrote a test like this in the past but I would like to get rid of it. Even testing locally I missed the test condition and locked myself out again (the task state mutation came too late, test was hanging.) It also requires a nanny + worker introspection which is quite unreadable. The only thing speaking for this proc suspension test is that it is the farthest away from the implementation around provoking the condition.

runtimes
```
1.56s call     distributed/tests/test_worker.py::test_gather_dep_exception_one_task_proc_suspend
0.26s call     distributed/tests/test_worker.py::test_gather_dep_exception_one_task_mock
0.26s call     distributed/tests/test_worker.py::test_gather_dep_exception_one_task_event_locked_comm
```

Reply:
In case I wasn't clear: I would like to use only one version of the tests but would like to have an agreement on how we want to test these things in the future

I don't have a strong opinion and can go either way. My preferences in declining order

* Event+Queue in utils_test
* mock
* proc suspension

For the record, I don't think this would be easily testable even if we were to refactor everything. sometimes refactoring makes testing easier but in this case it's one of these network/latencies which are hard to encapsulate. This is also why I'm slightly favouring the event+queue approach

Any opinions welcome!